src/encode/n_subst.ml

```diff
@@ -0,0 +1,256 @@
+and app_ix s n = match s with
+  | Shift m -> Ix (m + n.core) @@ n
+  | Cons (op, _) when n.core = 1 -> op.core @@ n
+  | Cons (_, s) -> app_ix s (n.core - 1 @@ n)
+  | Compose (ss, tt) -> app_expr ss (app_ix tt n)
+  | Bump (k, ss) when n.core <= k -> Ix n.core @@ n
+  | Bump (k, ss) -> app_expr (Shift k) (app_ix ss (n.core - k @@ n))
+
+and app_defn s d =
+  { d with core = app_defn_ s d.core }
+
+and app_defn_ s = function
+  | Recursive (nm, shp) -> Recursive (nm, shp)
+  | Operator (nm, e) -> Operator (nm, app_expr s e)
+  | Instance (nm, inst) -> Instance (nm, app_instance s inst)
+  | Bpragma (nm, e, l) -> Bpragma (nm, app_expr s e, l)
+
+and app_defns s = function
+  | [] -> (s, [])
+  | d :: ds ->
+      let d = app_defn s d in
+      let (s, ds) = app_defns (bump s) ds in
+      (s, d :: ds)
+
+and app_instance s i =
+  let s = bumpn (List.length i.inst_args) s in
+  { i with inst_sub = List.map (fun (v, e) -> (v, app_expr s e)) i.inst_sub }
+
+and app_sequent s sq =
+  let (s, cx) = app_hyps s sq.context in
+  { context = cx ; active = app_expr s sq.active }
+
+and app_hyps s cs = match Deque.front cs with
+  | None -> (s, Deque.empty)
+  | Some (c, cs) ->
+      let c = app_hyp s c in
+      let (s, cs) = app_hyps (bump s) cs in
+      (s, Deque.cons c cs)
+
+and app_hyp s h = match h.core with
+  | Fresh (x, shp, lv, b) -> Fresh (x, shp, lv, app_dom s b) @@ h
+  | Flex v -> Flex v @@ h
+  | Defn (d, wd, us, ex) -> Defn (app_defn s d, wd, us, ex) @@ h
+  | Fact (e, us,tm) -> Fact (app_expr s e, us,tm) @@ h
+
+and app_dom s = function
+  | Unbounded -> Unbounded
+  | Bounded (dom, vis) -> Bounded (app_expr s dom, vis)
+
+let rec app_spine s = function
+  | [] -> []
+  | e :: es ->
+      app_expr s e :: app_spine (bump s) es
+
+open Format
+open Expr.Fmt
+
+let pp_print_sub cx ff =
+  let rec handle ff s = match s with
+    | Shift n ->
+        fprintf ff "^%d" n
+    | Cons (op, s) ->
+        pp_print_expr cx ff op ;
+        pp_print_string ff " ." ;
+        pp_print_space ff () ;
+        handle ff s
+    | Compose (s, t) ->
+        fprintf ff "(%a) o (%a)" handle s handle t
+    | Bump (n, s) ->
+        fprintf ff "%d(%a)" n handle s
+  in
+  fprintf ff "@[<b1>[%a]@]" handle
+
+
+let extract sq k =
+  let prefix : hyp Deque.dq ref = ref Deque.empty in
+  let rec scan n hs = match Deque.front hs with
+    | None -> failwith "scan"
+    | Some (h, hs) ->
+        if n = k then (h, hs) else
+          begin
+            prefix := Deque.snoc !prefix h ;
+            scan (n + 1) hs
+          end
+  in try begin
+    let (h, hs) = scan 0 sq.context in
+    let e = match h.core with
+      | Fact (e, _, _) -> app_expr (shift (Deque.size hs)) e
+      | _ -> failwith "exprtract" in
+    let (s, hs) = app_hyps (shift (-1)) hs in
+    let sq = {
+      context = Deque.append !prefix hs ;
+      active = app_expr s sq.active ;
+    } in
+    Some (sq, e)
+  end with _ -> None
```

Comment:
Sorry, this is not actually my code but a copy-paste of the Subst module. I know this is dirty, but I needed to alter it for the encoding without breaking TLAPS.

Since then someone wrote a class for applying substitutions, and I can just use inheritance to solve my problem. This whole file is going to change.